internal/util/function/minhash_function.go

```diff
@@ -0,0 +1,664 @@
+#include <stdlib.h>
+#include "segcore/minhash_c.h"
+#include "segcore/tokenizer_c.h"
+*/
+import "C"
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"unsafe"
+
+	"github.com/cockroachdb/errors"
+	"github.com/milvus-io/milvus-proto/go-api/v2/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus/internal/util/analyzer"
+	"github.com/milvus-io/milvus/internal/util/analyzer/canalyzer"
+)
+
+// MinHashFunctionRunner
+// Input: string (text)
+// Output: []byte (binary vector - MinHash signature vector)
+const (
+	// outter  parameters
+	NumHashesKey   = "num_hashes"
+	ShingleSizeKey = "shingle_size"
+	HashFuncKey    = "hash_function"
+	TokenLevelKey  = "token_level" // "char" for character-level n-grams, "word" for word-level (default)
+	SeedKey        = "seed"
+	// internal  parameters
+	PermutationsKey    = "permutations" //  Serialized (a,b) parameters
+	defaultShingleSize = 3
+	defaultSeed        = 1
+	// const value
+	// Mersenne prime for universal hashing: 2^61 - 1
+	// Using a large prime to reduce collisions
+	mersennePrime = (1 << 61) - 1
+	maxHash32     = 0xFFFFFFFF
+)
+
+// HashFunction type
+type HashFunction int
+
+const (
+	// todo: support more hash functions
+	HashFuncSHA1 HashFunction = iota
+	HashFuncXXHash64
+)
+
+type MinHashFunctionRunner struct {
+	mu     sync.RWMutex
+	closed bool
+
+	tokenizer   analyzer.Analyzer // word-level tokenizer
+	funSchema   *schemapb.FunctionSchema
+	inputField  *schemapb.FieldSchema
+	outputField *schemapb.FieldSchema
+
+	// MinHash specific parameters
+	numHashes    int          // MinHash signature vector dimension
+	shingleSize  int          // N-gram, N size
+	hashFunc     HashFunction // Hash function to use
+	useCharToken bool         // true: character-level n-grams, false: word-level tokens + shingles
+
+	// Universal hash family parameters: h(x) = ((a * x + b) mod p) mod m
+	// Each permutation has its own (a, b) pair
+	permA []uint64 // 'a' (must be odd for full period)
+	permB []uint64 // 'b'
+
+	// Buffer pool for reducing allocations
+	bufferPool sync.Pool
+}
+
+func NewMinHashFunctionRunner(
+	collSchema *schemapb.CollectionSchema,
+	funSchema *schemapb.FunctionSchema,
+) (FunctionRunner, error) {
+	if len(funSchema.GetOutputFieldIds()) != 1 {
+		return nil, fmt.Errorf("minhash function should only have one output field, but now %d", len(funSchema.GetOutputFieldIds()))
+	}
+	var inputField, outputField *schemapb.FieldSchema
+	for _, field := range collSchema.GetFields() {
+		if field.GetFieldID() == funSchema.GetOutputFieldIds()[0] {
+			outputField = field
+		}
+
+		if field.GetFieldID() == funSchema.GetInputFieldIds()[0] {
+			inputField = field
+		}
+	}
+	if outputField == nil {
+		return nil, errors.New("no output field")
+	}
+
+	params := getAnalyzerParams(inputField)
+	tokenizer, err := analyzer.NewAnalyzer(params)
+	if err != nil {
+		return nil, err
+	}
+
+	var numHashes int
```

Comment:
If the Param doesn’t have a value set here, there should be a default value, right?